pkg/operator/controllers/imageconfig/image_controller.go

```diff
@@ -0,0 +1,131 @@
+package imageconfig
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+)
+
+const imageConfigResource = "cluster"
+
+type Reconciler struct {
+	log       *logrus.Entry
+	arocli    aroclient.Interface
+	configcli configclient.Interface
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, configcli configclient.Interface) *Reconciler {
+	return &Reconciler{
+		log:       log,
+		arocli:    arocli,
+		configcli: configcli,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	// Get cluster
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Feature flag
+	if !instance.Spec.Features.ReconcileImageConfig {
+		return reconcile.Result{}, nil
+	}
+
+	// Check for cloud type
+	requiredRegistries, err := getCloudAwareRegistries(instance)
+	if err != nil {
+		return reconcile.Result{Requeue: false}, err
+	}
+
+	// Get image.config yaml
+	imageconfig, err := r.configcli.ConfigV1().Images().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Append to allowed registries
+	if imageconfig.Spec.RegistrySources.AllowedRegistries != nil {
```

Comment:
According to the documentation it is either Allowed or Blocked. It cannot be both. https://docs.openshift.com/container-platform/4.7/openshift_images/image-configuration.html#images-configuration-parameters_image-configuration 

I believe if both of these are set, the code should fail fast and log error properly, so we are aware of the issue and can act upon it. Unfortunately we cannot convert one to the other in simple manner and if customer set it, he had an intention for it.

Would you please update the code to cover this edge case?

Reply:
Thank you for the review @petrkotas ! to be honest, my first iteration of code looked very different from current code.. so its a big kudos to our review process. I will add the changes you suggested, along with another issue i found where controller adds registry even if it already exists.